symbolic/src/Data/Macaw/Symbolic/Stack.hs

```diff
@@ -0,0 +1,115 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE ImplicitParams #-}
+{-# LANGUAGE ImportQualifiedPost #-}
+{-# LANGUAGE TypeOperators #-}
+
+module Data.Macaw.Symbolic.Stack
+  ( stackArray
+  , mallocStack
+  , ExtraStackSlots(..)
+  , ArrayStack(..)
+  , createArrayStack
+  ) where
+
+import Data.BitVector.Sized qualified as BVS
+import Data.Parameterized.Context as Ctx
+import Lang.Crucible.Backend qualified as C
+import Lang.Crucible.LLVM.DataLayout qualified as CLD
+import Lang.Crucible.LLVM.MemModel qualified as CLM
+import What4.Interface qualified as WI
+import What4.Symbol qualified as WSym
+
+-- | Helper, not exported
+stackAlign :: CLD.Alignment
+stackAlign = CLD.noAlignment
+
+-- | Create an SMT array representing the program stack.
+--
+-- Builds a 'WI.freshConstant' with name "stack_array".
+stackArray ::
+  (1 WI.<= w) =>
+  CLM.HasPtrWidth w =>
+  WI.IsSymExprBuilder sym =>
+  sym ->
+  IO (WI.SymArray sym (Ctx.SingleCtx (WI.BaseBVType w)) (WI.BaseBVType 8))
+stackArray sym =
+  WI.freshConstant
+    sym
+    (WSym.safeSymbol "stack_array")
+    (WI.BaseArrayRepr (Ctx.singleton (WI.BaseBVRepr ?ptrWidth)) WI.knownRepr)
+
+-- | Create an allocation representing the program stack, using 'CLM.doMalloc'.
+--
+-- This allocation is:
+--
+-- * mutable, because the program must write to the stack
+-- * of kind 'CLM.StackAlloc'
+-- * without alignment ('CLD.noAlignment')
+mallocStack ::
+  C.IsSymBackend sym bak =>
+  CLM.HasPtrWidth w =>
+  (?memOpts :: CLM.MemOptions) =>
+  CLM.HasLLVMAnn sym =>
+  bak ->
+  CLM.MemImpl sym ->
+  -- | Size of stack allocation
+  WI.SymExpr sym (WI.BaseBVType w) ->
+  IO (CLM.LLVMPtr sym w, CLM.MemImpl sym)
+mallocStack bak mem sz =
+  CLM.doMalloc bak CLM.StackAlloc CLM.Mutable "stack_alloc" mem sz stackAlign
+
+-- | Allocate this many pointer-sized stack slots, which are reserved for
+-- stack-spilled arguments.
+newtype ExtraStackSlots = ExtraStackSlots { getExtraStackSlots :: Int }
+  -- Derive the Read instance using the `newtype` strategy, not the
+  -- `stock` strategy. This allows parsing ExtraStackSlots values in
+  -- optparse-applicative-based command-line parsers using the `Read` instance.
+  deriving newtype (Enum, Eq, Integral, Num, Ord, Read, Real, Show)
+
+-- | An allocartion representing the program stack, backed by an SMT array
+data ArrayStack sym w
+  = ArrayStack
+    { -- | Pointer to the base of the stack array
+      arrayStackBasePtr :: CLM.LLVMPtr sym w
+      -- | Pointer to the top (end) of the stack array
+    , arrayStackTopPtr :: CLM.LLVMPtr sym w
+      -- | SMT array backing the stack allocation
+    , arrayStackVal :: WI.SymArray sym (Ctx.SingleCtx (WI.BaseBVType w)) (WI.BaseBVType 8)
```

Comment:
Do you foresee it being convenient to have different underlying stack representations besides an SMT array? (Perhaps that's not in scope for this PR, but I thought I'd bring it up given your comments in https://github.com/GaloisInc/macaw/issues/430#issuecomment-2318712835.)

Reply:
I think it *might* be the case that the SMT array memory model can't detect uninitialized reads? I might be misremembering that, though. Frankly, I don't understand the trade-offs between the models, so I can't really comment on whether or not it would be useful to have both representations. In either case, doesn't hurt to be explicit here.